### PR TITLE
(BOLT-608) Add `local` transport for Windows OS

### PR DESCRIPTION
**Pull request by donoghuc**
Add support for local transport on windows OS.

### CODE REVIEW

File: lib/bolt/transport/local.rb [hunk 1]
```diff
@@ -65,27 +60,298 @@ def run_command(target, command, _options = {})
...
+          if arg =~ / /
+            "\"#{arg}\""
+          else
+            arg
+          end
+        end
+      end
+
+      def shell_init
+        command = <<-PS
```
Comment (MikaelSmith) [inline lib/bolt/transport/local.rb]: I'm guessing this is copied from WinRM. Can we pull shared knowledge around how to run things in powershell out into a separate class that the WinRM and Local transports can share?

File: Rakefile [hunk 2]
```diff
@@ -127,6 +127,11 @@ namespace :integration do
     t.rspec_opts = '--tag bash'
   end
 
+  desc 'Run tests that require windows local host'
+  RSpec::Core::RakeTask.new(:bash) do |t|
```
Comment (MikaelSmith) [inline Rakefile]: This should have a different name. I'd opt for `:powershell`.

File: spec/integration/local_spec.rb [hunk 3]
```diff
@@ -41,4 +41,22 @@
       expect(result[0]['_error']).to be
     end
   end
+
+  context 'when using CLI options on POSIX OS', windows: true do
```
Comment (MikaelSmith) [inline spec/integration/local_spec.rb]: I like the naming to be closer to the context (i.e. powershell vs bash).

File: lib/bolt/transport/local.rb (hunk 1 above)
Comment (nicklewis) [reply lib/bolt/transport/local.rb]: That would also be useful if we eventually support powershell over SSH. Like a transport vs. runtime separation

Comment (donoghuc) [reply lib/bolt/transport/local.rb]: Will absolutely refactor to use a shared class for powershell code. I was just trying to get something working to write tests against. Having tests will help me with the refactor.

File: spec/integration/local_spec.rb [hunk 4]
```diff
@@ -41,4 +51,68 @@
...
+      end
+    end
+
+    it 'runs a task reading from stdin', :reset_puppet_settings do
+      result = run_one_node(%w[task run sample::winstdin message=µsomemessage] + config_flags)
+      output = result['_output'].strip
+      output = output.force_encoding(Encoding.default_external).encode!(''.encoding) if RUBY_VERSION =~ /^2\.3\./
+      expect(output).to match(/STDIN: {"message":"µsomemessage"/)
+    end
+
```
[the remark this answers is not in the document]
Comment (donoghuc) [reply spec/integration/local_spec.rb]: I think this is a bug outside of this work. For example from a windows machine: 
```
C:\Users\Administrator\bolt>bundle exec bolt task run sample::winstdin message=µsomemessage --modulepath .\spec\fixtures\modules --transport winrm --user Administrator --password *** --no-ssl -n fwviuuk1typ4xz4.delivery.puppetlabs.net
bundle exec bolt task run sample::winstdin message=µsomemessage --modulepath .\spec\fixtures\modules --transport winrm --user Administrator --password Qu@lity! --no-ssl -n fwviuuk1typ4xz4.delivery.puppetlabs.net
Started on fwviuuk1typ4xz4.delivery.puppetlabs.net...
Finished on fwviuuk1typ4xz4.delivery.puppetlabs.net:
  STDIN: {"message":"-¦somemessage","_task":"sample::winstdin"}
  {
  }
Successful on 1 node: fwviuuk1typ4xz4.delivery.puppetlabs.net
Ran on 1 node in 2.36 seconds
```
I think it is a problem outside of bolt: 
```
PS C:\Users\Administrator\bolt> echo "echo µsomemessage;" | powershell -noprofile -noninteractive -command '$input | iex'
echo "echo µsomemessage;" | powershell -noprofile -noninteractive -command '$input | iex'
??somemessage
```
```
PS C:\Users\Administrator\bolt> echo "echo somemessage;" | powershell -noprofile -noninteractive -command '$input | iex'
echo "echo somemessage;" | powershell -noprofile -noninteractive -command '$input | iex'
somemessage
```

File: Rakefile [hunk 5]
```diff
@@ -24,7 +24,7 @@ end
 
 desc "Run RSpec tests for TravisCI that don't require WinRM"
 RSpec::Core::RakeTask.new(:travisci) do |t|
-  t.rspec_opts = '--tag ~winrm --tag ~appveyor_agents --tag ~puppetserver'
+  t.rspec_opts = '--tag ~winrm --tag ~appveyor_agents --tag ~puppetserver --tag ~powershell'
```
Comment (adreyer) [inline Rakefile]: Are you running the powershell tests on linux?

**Comment by donoghuc** [comment]
I have a testing environment set up at `jbwztop5z8no6nc.delivery.puppetlabs.net` feel free to use it. See example below. 

```
C:\Users\Administrator\bolt>bundle exec bolt script run test.rb one -n localhost
bundle exec bolt script run test.rb one -n localhost
Started on localhost...
Finished on localhost:
  STDOUT:
    Ruby script
    one
Successful on 1 node: localhost
Ran on 1 node in 0.81 seconds
```

File: lib/bolt/transport/winrm.rb [hunk 6]
```diff
@@ -78,30 +75,11 @@ def run_script(target, script, arguments, _options = {})
...
-{
-  Invoke-Command @invokeArgs
-}
-catch
-{
-  Write-Error $_.Exception
-  exit 1
-}
-          PS
+            if Bolt::Transport::Powershell.powershell_file?(remote_path)
```
Comment (adreyer) [inline lib/bolt/transport/winrm.rb]: I think include the module so you can just call these methods

File: Rakefile (hunk 5 above)
Comment (MikaelSmith) [reply Rakefile]: My suggestion was the opposite. We might support powershell on linux, and there's probably nothing windows-specific about it. https://github.com/puppetlabs/bolt/pull/775#discussion_r238873814 specifically.

Comment (donoghuc) [reply Rakefile]: I have not been running the powershell tests on my laptop. In fact I had not even really considered that as an option for linux. Will explore how this should work more in depth.

The following work with PR as is.
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ bolt command run "powershell -noprofile -noninteractive -command 'Write-host hi'" -n localhost
Started on localhost...
Finished on localhost:
  STDOUT:
    hi
Successful on 1 node: localhost
Ran on 1 node in 0.89 seconds
```
```
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ cat argv.ps1 
#!/usr/bin/powershell
Write-Host $args
cas@cas-ThinkPad-T460p:~/working_dir/bolt$ bolt script run argv.ps1 one -n localhost
Started on localhost...
Finished on localhost:
  STDOUT:
    one
Successful on 1 node: localhost
Ran on 1 node in 1.21 seconds
```

**Comment by tphoney** [comment]
Looking forward to seeing this. +1 to all involved.

File: spec/integration/local_spec.rb [hunk 7]
```diff
@@ -4,22 +4,22 @@
...
 
-  let(:whoami) { "whoami" }
   let(:modulepath) { File.join(__dir__, '../fixtures/modules') }
-  let(:stdin_task) { "sample::stdin" }
   let(:uri) { 'localhost,local://foo' }
   let(:user) { ENV['USER'] }
 
   after(:each) { Puppet.settings.send(:clear_everything_for_tests) }
 
-  context 'when using CLI options' do
```
Comment (MikaelSmith) [inline spec/integration/local_spec.rb]: Are there common tests we can run on all platforms?

File: lib/bolt/transport/local.rb [hunk 8]
```diff
@@ -73,21 +69,34 @@ def run_script(target, script, arguments, _options = {})
 
           # unpack any Sensitive data AFTER we log
           arguments = unwrap_sensitive_args(arguments)
-          if arguments.empty?
-            # We will always provide separated arguments, so work-around Open3's handling of a single
-            # argument as the entire command string for script paths containing spaces.
-            arguments = ['']
+          if Bolt::Util.windows?
```
Comment (MikaelSmith) [inline lib/bolt/transport/local.rb]: Seems promising.

File: lib/bolt/transport/winrm.rb (hunk 6 above)
Comment (donoghuc) [reply lib/bolt/transport/winrm.rb]: How about just changing `Bolt::Transport::Powershell.` to `Powershell.`? Seems like this solution cuts down on characters but helps disambiguate methods like `run_script`.

File: lib/bolt/transport/local.rb [hunk 9]
```diff
@@ -111,13 +120,46 @@ def run_task(target, task, arguments, _options = {})
           # unpack any Sensitive data, write it to a separate variable because
           # we log 'arguments' below
           unwrapped_arguments = unwrap_sensitive_args(arguments)
-          stdin = STDIN_METHODS.include?(input_method) ? JSON.dump(unwrapped_arguments) : nil
-          env = ENVIRONMENT_METHODS.include?(input_method) ? envify_params(unwrapped_arguments) : nil
 
+          if Bolt::Util.windows?
+            # WINDOWS
+            stdin = if STDIN_METHODS.include?(input_method)
```
Comment (MikaelSmith) [inline lib/bolt/transport/local.rb]: `stdin` is constructed identically here. Would be nice not to duplicate it.

File: lib/bolt/transport/local.rb [hunk 10]
```diff
@@ -111,13 +120,46 @@ def run_task(target, task, arguments, _options = {})
...
+              end
+          else
+            # POSIX
+            stdin = STDIN_METHODS.include?(input_method) ? JSON.dump(unwrapped_arguments) : nil
+            env = ENVIRONMENT_METHODS.include?(input_method) ? envify_params(unwrapped_arguments) : nil
+
+            output = @conn.execute(script, stdin: stdin, env: env, dir: dir)
+          end
           # log the arguments with sensitive data redacted, do NOT log unwrapped_arguments
           logger.debug("Running '#{script}' with #{arguments}")
```
Comment (MikaelSmith) [inline lib/bolt/transport/local.rb]: This logging should happen before the task is executed.

File: spec/integration/local_spec.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (MikaelSmith) [reply spec/integration/local_spec.rb]: What version of Ruby are you using?

**Comment by MikaelSmith** [comment]
Needs a rebase.

File: lib/bolt/inventory.rb [hunk 11]
```diff
@@ -174,7 +174,7 @@ def update_target(target)
         data['config'] = {}
       end
 
-      unless Bolt::Util.windows? || data['config']['transport']
+      unless data['config']['transport']
         data['config']['transport'] = 'local' if target.name == 'localhost'
```
Comment (MikaelSmith) [inline lib/bolt/inventory.rb]: Whole thing could be collapsed to `data['config']['transport'] ||= 'local' if target.name == 'localhost'`.

File: lib/bolt/transport/local.rb [hunk 12]
```diff
@@ -108,16 +121,43 @@ def run_task(target, task, arguments, _options = {})
...
 
-          output = @conn.execute(script, stdin: stdin, env: env, dir: dir)
+            output =
+              if Powershell.powershell_file?(script) && stdin.nil?
+                command = Powershell.run_ps_task(arguments, script, input_method)
+                command = environment_params + Powershell.shell_init + command
+                if input_method == 'powershell'
+                  @conn.execute(command, dir: dir, env: "powershell.exe")
+                else
+                  @conn.execute(command, dir: dir, stdin: stdin, env: "powershell.exe")
```
Comment (MikaelSmith) [inline lib/bolt/transport/local.rb]: This feels like a weird use of `env`. `env` was supposed to be about setting up environment variables. However it looks like the way we order that in shell.rb doesn't work for Windows.

**Pull request merged**